src/main/java/core/basesyntax/MyLinkedList.java

```
@@ -3,45 +3,191 @@
+        Node<T> newNode = new Node<>(value);
+        if (size == 0) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+
+        Node<T> newNode = new Node<>(value);
+
+        if (size == 0) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        } else {
+            Node<T> prevNode = head;
+            for (int i = 0; i < index - 1; i++) {
+                prevNode = prevNode.next;
+            }
+            newNode.next = prevNode.next;
+            newNode.prev = prevNode;
+            prevNode.next.prev = newNode;
+            prevNode.next = newNode;
+        }
+
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        if (index == 0) {
+            return head.data;
+        } else if (index == size - 1) {
+            return tail.data;
+        } else {
+            int counter = 0;
+            Node<T> temp = head;
+            while (temp != null) {
+                if (counter == index) {
+                    return temp.data;
+                }
+                temp = temp.next;
+                counter++;
+            }
+        }
         return null;
```

Issue: The `get(int index)` method returns `null` at the end if the node is not found. According to the checklist, 'The method should throw an IndexOutOfBoundsException if the index is invalid.' Since you already check for invalid indices at the start, this return statement should never be reached. However, for clarity and to avoid hiding potential bugs, it is better to throw an exception or refactor the method to guarantee a return value for all valid indices. Consider removing the final `return null;` and restructuring the method to always return the found node's data.